File: src/main/java/bankaccount/history/HistoryLine.java
package bankaccount.history;

import bankaccount.Amount;
import bankaccount.OperationType;
import bankaccount.Printer;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public abstract class HistoryLine {
    private static final String SEPARATOR = " | ";
    private static final String CURRENCY = "€";
    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("dd/MM/YYYY");

    protected final OperationType operationType;
    protected final Amount amount;
    protected final LocalDate date;
    protected final Amount currentBalance;

    protected HistoryLine(OperationType operationType, Amount amount, LocalDate date, Amount currentBalance) {
        this.operationType = operationType;
        this.amount = amount;
        this.date = date;
        this.currentBalance = currentBalance;
    }

    public void print(Printer printer) {
        String lineRepresentation = operationType.toString()
                + SEPARATOR + FORMATTER.format(date)
                + SEPARATOR + amount.print() + CURRENCY
                + SEPARATOR + currentBalance.print() + CURRENCY;

        printer.print(lineRepresentation);
    }

    public abstract Amount combineAmounts(Amount amount);
}

File: src/main/kotlin/bankaccount/bankaccount/HistoryLine.kt
package bankaccount.bankaccount

import java.time.LocalDate
import java.time.format.DateTimeFormatter

abstract class HistoryLine(
    protected val amount: Amount,
    protected val date: LocalDate,
    protected val currentBalance: Amount
) {
    fun print(printer: Printer) {
        val lineRepresentation = (printOperationType()
                + SEPARATOR + FORMATTER.format(date)
                + SEPARATOR + amount.print()
                + SEPARATOR + currentBalance.print())
        printer.print(lineRepresentation)
    }

    protected abstract fun printOperationType(): String
    abstract fun combineAmounts(amount: Amount): Amount

    companion object {
        private const val SEPARATOR = " | "
        private val FORMATTER = DateTimeFormatter.ofPattern("dd/MM/YYYY")
    }
}
